### ISSUE TITLE
Port SMP x86

### ISSUE THREAD

**Issue by luigidcsoares** [nanvix/nanvix#254]
# Port SMP for x86 architecture

## Description 
There are some initial steps to walk through before we can achieve this feature. At first, we need to search for a data structure named [RSDP](https://wiki.osdev.org/RSDP) (Root System Description Pointer), which is used in the [ACPI](https://wiki.osdev.org/ACPI) programming interface. To do that, one may need to look at two different places: the first 1 KB of the [EBDA](https://wiki.osdev.org/EBDA#Extended_BIOS_Data_Area_.28EBDA.29) (Extendes BIOS Data Area), or in the memory region from 0x000E0000 to 0x000FFFF (the main [BIOS](https://wiki.osdev.org/BIOS) area below 1 MB). All the instructions you need to do that can be find at [OSDev](https://wiki.osdev.org).

RSDP contains a 32-bit address to a table named [RSDT](https://wiki.osdev.org/RSDT) (Root System Descriptor Table) that we'll need to parse. It may also contains a 64-bit address to a table named [XSDT](https://wiki.osdev.org/XSDT) (eXtended System Descriptor Table), that is just the 64-bit version of the ACIP RSDT. RSDT give us pointers to various structures (System Descriptor Tables). The one we need is named [MADT](https://wiki.osdev.org/MADT) (Multiple APIC Description Table), that describes how the APIC works.

MADT contains a lot of entries of different types. Type 0 is the one we're interested, because it describes the processors local APIC. It contains three fields: ACPI Processor ID, APIC ID and a flag that says whether that processor is enabled or not. We'll use those APICs to wake up the cores sending specific IPIs.

## What is already done?
Everything related to finding and parsing RSDP, RSDT and MADT is already done, except for looking at EBDA that couldn't be tested properly. All those things were implemented in assembly, but they have their structures defined in their own headers, which may be useful. Also, there is a checksum function implemented in assembly that should work with all those tables.

## References
[OSDev](https://wiki.osdev.org/Expanded_Main_Page) is your friend here. You should find almost everything you need in those pages. Here are some useful links to look:

  * [ACPI](https://wiki.osdev.org/ACPI)
  * [RSDP](https://wiki.osdev.org/RSDP)
  * [RSDT](https://wiki.osdev.org/RSDT)
  * [MADT](https://wiki.osdev.org/MADT)
 
[ACPI Specification](http://www.uefi.org/sites/default/files/resources/ACPI%206_2_A_Sept29.pdf) could be useful to check specific things related to all those tables and structures. This is the standard documentation and you should spend some time here.

[MADT example](https://github.com/seL4/libplatsupport/blob/master/plat_include/pc99/platsupport/plat/acpi/tables/madt.h) is an example of how to implement the MADT structure.

[ACPI example](https://chromium.googlesource.com/chromiumos/third_party/coreboot/+/firmware-stout-2817.B/src/arch/x86/boot/acpi.c) is an example of how to implement ACPI.

You should also look at the answer on the following link, which contains informations about how to actually start the another cores: [How to use APIC to create IPIs to wake the APs for SMP in x86 assembly?](https://stackoverflow.com/questions/16364817/how-to-use-the-apic-to-create-ipis-to-wake-the-aps-for-smp-in-x86-assembly)
